acceptance/tests/pxp-module-puppet/run_puppet_killed_puppet.rb

```diff
@@ -0,0 +1,132 @@
+require 'pxp-agent/test_helper.rb'
+require 'pxp-agent/config_helper.rb'
+require 'puppet/acceptance/environment_utils'
+
+STATUS_QUERY_MAX_RETRIES = 60
+STATUS_QUERY_INTERVAL_SECONDS = 1
+
+test_name 'Run Puppet while a Puppet Agent run is in-progress, wait for it to be killed' do
+  extend Puppet::Acceptance::EnvironmentUtils
+
+  env_name = test_file_name = File.basename(__FILE__, '.*')
+  environment_name = mk_tmp_environment(env_name)
+
+  step 'On master, create a new environment that will result in a slow run' do
+    site_manifest = "#{environmentpath}/#{environment_name}/manifests/site.pp"
+    create_remote_file(master, site_manifest, <<-SITEPP)
+node default {
+  case $::osfamily {
+    'windows': { exec { 'sleep':
+                        command => 'true',
+                        unless  => 'sleep 10', #PUP-5806
+                        path    => 'C:\\cygwin64\\bin',} }
+    default:  { exec { '/bin/sleep 30': } }
+  }
+}
+SITEPP
+    on(master, "chmod 644 #{site_manifest}")
+  end
+
+  step 'Ensure each agent host has pxp-agent running and associated' do
+    agents.each do |agent|
+      on agent, puppet('resource service pxp-agent ensure=stopped')
+      create_remote_file(agent, pxp_agent_config_file(agent), pxp_config_json_using_puppet_certs(master, agent).to_s)
+      on agent, puppet('resource service pxp-agent ensure=running')
+      show_pcp_logs_on_failure do
+        assert(is_associated?(master, "pcp://#{agent}/agent"),
+               "Agent #{agent} with PCP identity pcp://#{agent}/agent should be associated with pcp-broker")
+      end
+    end
+  end
+
+  lockfiles = []
+  agents.each do |agent|
+    step "Get lockfile location for #{agent}" do
+      on agent, puppet('agent', '--configprint', 'agent_catalog_run_lockfile') do |result|
+        lockfiles << result.stdout.chomp
+      end
+    end
+  end
+
+  agents.each do |agent|
+    step "Start a long-running Puppet agent job on #{agent}" do
+      on agent, puppet('agent', '--test', '--environment', environment_name, '--server', "#{master}",
+                       '>/dev/null & echo $!')
+    end
+  end
+
+  # TODO: switch to checking all agents after each sleep
+  agents.zip(lockfiles).each do |agent, lockfile|
+    step "Check for lockfile on #{agent}" do
+      lockfile_exists = false
+      for i in 1..50
```

Comment:
You could maybe use [retry_on](http://www.rubydoc.info/github/puppetlabs/beaker/Beaker%2FDSL%2FHelpers%2FHostHelpers%3Aretry_on) here to save writing a loop.
If using retry_on , it needs to a begin/rescue to translate anything it throws into a meaningful assertion failure.